### PR TITLE
New Resource: `azurerm_dynatrace_monitor`

### PR DESCRIPTION
**Pull request by jiaweitao001**

### CODE REVIEW

File: internal/services/dynatrace/client/client.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-azure-sdk/resource-manager/dynatrace/2023-04-27/monitors"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/common"
+)
+
+type Client struct {
+	MonitorClient *monitors.MonitorsClient
```
Comment (magodo) [inline internal/services/dynatrace/client/client.go:11]: Currently, for the client which is a thin wrapper of the underlying client, instead of having multiple underlying clients, we prefer defining it as an embedded field:

```suggestion
	*monitors.MonitorsClient
```

File: internal/services/serviceconnector/service_connector_app_service_resource_test.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ func (r ServiceConnectorAppServiceResource) Exists(ctx context.Context, client *
 		}
 		return nil, fmt.Errorf("retrieving %s: %+v", *id, err)
 	}
+
```
Comment (magodo) [inline internal/services/serviceconnector/service_connector_app_service_resource_test.go:35]: Can we revert this unrelated change?

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 3]
```diff
@@ -0,0 +1,116 @@
+---
+subcategory: "Dynatrace"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_dynatrace_monitors"
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:4]: Shall we rename it to its singular form: 
```suggestion
page_title: "Azure Resource Manager: azurerm_dynatrace_monitor"
```

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  name     = "example-resources"
+  location = "West Europe"
+}
+
+resource "azurerm_dynatrace_monitors" "example" {
+  name                            = "exmpledynatracemonitor"
+  resource_group_name             = azurerm_resource_group.example.name
+  location                        = azurerm_resource_group.test.location
+  identity_type                   = "SystemAssigned"
+  monitoring_status               = "Enabled"
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:27]: Can we rename it to `enabled` for short?

File: internal/services/dynatrace/helper.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
+package dynatrace
```
Comment (magodo) [inline internal/services/dynatrace/helper.go:1]: Is there a specific reason to put these in a separate file? Otherwise, can we put them back to the resource file, also, define the schemas in the root schema composite literally?

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    email        = "alice@microsoft.com"
+    phone_number = "123456"
+    country      = "westus"
+  }
+
+  plan {
+    usage_type     = "COMMITTED"
+    billing_cycle  = "MONTHLY"
+    plan           = "azureportalintegration_privatepreview@TIDgmz7xq9ge3py"
+    effective_date = "2019-08-30T15:14:33Z"
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource_test.go:112]: As this is the basic test, can we ensure all the optional properties are not specified?

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	return fmt.Sprintf(`
+%[1]s
+
+resource "azurerm_dynatrace_monitors" "test" {
+  name                     = "acctestacc%[2]d"
+  resource_group_name      = azurerm_resource_group.test.name
+  location                 = azurerm_resource_group.test.location
+  identity_type            = "SystemAssigned"
+  monitoring_enabled       = true
+  marketplace_subscription = "Active"
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource_test.go:133]: Can we change these to different values than basic?

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go (hunk 7 above)
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource_test.go:133]: Will do.

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+resource "azurerm_resource_group" "example" {
+  name     = "example-resources"
+  location = "West Europe"
+}
+
+resource "azurerm_dynatrace_monitors" "example" {
+  name                            = "exmpledynatracemonitor"
+  resource_group_name             = azurerm_resource_group.example.name
+  location                        = azurerm_resource_group.test.location
+  identity_type                   = "SystemAssigned"
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:26]: This also needs to be updated

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+---
+
+A `plan` block supports the following:
+
+* `billing_cycle` - (Optional) Different billing cycles. Possible values are `MONTHLY`, `WEEKLY`.
+
+* `effective_date` - (Required) Date when plan was applied.
+
+* `plan` - (Required) Plan id as published by Dynatrace.
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:83]: Can we put these twos at the top as they are required?

File: internal/services/dynatrace/dynatrace_monitors_resource.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+var _ sdk.ResourceWithUpdate = MonitorsResource{}
+
+type MonitorsResourceModel struct {
+	Name                          string                                     `tfschema:"name"`
+	ResourceGroup                 string                                     `tfschema:"resource_group_name"`
+	Location                      string                                     `tfschema:"location"`
+	MonitoringStatus              bool                                       `tfschema:"monitoring_enabled"`
+	MarketplaceSubscriptionStatus string                                     `tfschema:"marketplace_subscription"`
+	Identity                      []identity.ModelSystemAssignedUserAssigned `tfschema:"identity"`
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource.go:29]: Can we use `identity.ModelSystemAssigned` instead as it only supports SAI?

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go (hunk 7 above)
Comment (magodo) [reply internal/services/dynatrace/dynatrace_monitors_resource_test.go]: Can we also change the values in  `user`, `plan` blocks between tests?

**Review by magodo — CHANGES_REQUESTED** [review]

File: internal/services/dynatrace/dynatrace_monitors_resource.go [hunk 11]
```diff
@@ -270,33 +272,17 @@ func expandDynatraceIdentity(input []identity.ModelSystemAssignedUserAssigned) (
...
-	for v := range pointer.From(input.UserAssignedIdentities) {
-		ids = append(ids, v)
-	}
-
-	return []identity.ModelSystemAssignedUserAssigned{
+	return []identity.ModelSystemAssigned{
 		{
-			Type:        identity.Type(input.Type),
-			IdentityIds: ids,
+			Type: identity.Type(input.Type),
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource.go:285]: Isn't the `PrincipalId` and `TenantId` are also needed to be flattend here (also need be defined in the schema)?

**Review by magodo — CHANGES_REQUESTED** [review]

File: internal/services/dynatrace/dynatrace_monitors_resource.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+					return fmt.Errorf("updating %s: %+v", *id, err)
+				}
+			}
+
+			return nil
+		},
+	}
+}
+
+func expandDynatraceIdentity(input []identity.ModelSystemAssigned) (*monitors.IdentityProperties, error) {
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource.go:265]: From my observation, this is different because in Swagger the definition of the identity properties is slightly different from the standard identity.

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go (hunk 7 above)
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource_test.go]: Dynatrace now only support update on `tags`.

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
+package dynatrace_test
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource_test.go:1]: Can we add a Precheck for all these test cases, as they are expensive to test (the [price](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/dynatrace.dynatrace_portal_integration?tab=PlansAndPrice))? The Precheck can be a check on a specific environment variable, that will skip the test case when that variable is not specified explicitly.

File: internal/services/dynatrace/dynatrace_monitors_resource.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+			Default:  true,
+		},
+
+		"marketplace_subscription": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				"Active",
+				"Suspended",
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource.go:75]: Is there a valid use case that the `marketplace_subscription` can be set to `Suspended`?

File: internal/services/dynatrace/dynatrace_monitors_resource.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+		},
+
+		"resource_group_name": commonschema.ResourceGroupName(),
+
+		"location": commonschema.Location(),
+
+		"monitoring_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
```
Comment (magodo) [inline internal/services/dynatrace/dynatrace_monitors_resource.go:65]: In most cases, this shall be enabled (as the resource is a monitor resource) when user wants to create it? Also as it is marked as `ForceNew: true`, does it make sense that we just remove this attribute and always set it as `true` when user wants to create it?

Otherwise, if this is togglable, then we may still keep this, but remove the `ForceNew: true`.

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+* `type` - (Required) The type of identity used for the resource. Only possible value is `SystemAssigned`.
+
+---
+
+A `plan` block supports the following:
+
+* `billing_cycle` - (Optional) Different billing cycles. Possible values are `MONTHLY`, `WEEKLY`.
+
+* `effective_date` - (Required) Date when plan was applied.
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:84]: From Portal, users don't need to specify this. As the document stated, it more looks like a computed value where the dynatrace env is created. So can we figure this out?

File: website/docs/r/dynatrace_monitors.html.markdown [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+* `plan` - (Required) Plan id as published by Dynatrace.
```
Comment (magodo) [inline website/docs/r/dynatrace_monitors.html.markdown:86]: Is there a source of the possible values of the `plan`? The one used in the acctest seems not something users can figure out?

**Review by magodo — CHANGES_REQUESTED** [review]

**Comment by jackofallops** [comment]
As discussed offline, closing this PR for now due to testing concerns. Once addressed we can look to re-open.

File: internal/services/dynatrace/dynatrace_monitors_resource_test.go (hunk 13 above)
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource_test.go:1]: Sure. Will add.

File: internal/services/dynatrace/dynatrace_monitors_resource.go (hunk 14 above)
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource.go:75]: I'm not so sure about the use cases, I can confirm with service team.

File: internal/services/dynatrace/dynatrace_monitors_resource.go (hunk 15 above)
Comment (jiaweitao001) [reply internal/services/dynatrace/dynatrace_monitors_resource.go:65]: This is not togglable for now. I will confirm with service team if it's ok to always set it as `true`.

File: website/docs/r/dynatrace_monitors.html.markdown (hunk 16 above)
Comment (jiaweitao001) [reply website/docs/r/dynatrace_monitors.html.markdown:84]: Will confirm with service team.
